File: pyRBDL/test/octave/mRBDL/Contact/DetectContact.m
function flag_contact = DetectContact(model, q, qdot, contact_cond)
%DetectContact - Detect contact by position and velocity of foot endpoint 
%
% Syntax: flag_contact = DetectContact(q, qdot)
%
% flag_contact: contact flag, 0-uncontact, 1-contact, 2-impact
    global ip;


    flag_contact = zeros(model.NC, 1);

    endpos = zeros(3, model.NC);
    endvel = zeros(3, model.NC);
    for i=1:model.NC
        % Calcualte pos and vel of foot endpoint, column vector
        endpos(:, i) = CalcBodyToBaseCoordinates(model, q, model.idcontact(i), model.contactpoint{i});
        endvel(:, i) = CalcPointVelocity(model, q, qdot, model.idcontact(i), model.contactpoint{i});
        
        % Detect contact
        flag_contact(i) = DeterminContactType(endpos(:, i), endvel(:, i), contact_cond);
    end
%     [endpos(:,1)',endvel(:,1)', flag_contact(1), 99 , endpos(:,2)',endvel(:,2)', flag_contact(2)]

    %% Determin the contact type 
    function flag = DeterminContactType(pos, vel, contact_cond)
        if pos(3) < contact_cond.contact_pos_lb(3)
            if vel(3) < contact_cond.contact_vel_lb(3)
                flag = 2; % impact
            else 
                if vel(3) > contact_cond.contact_vel_ub(3)
                    flag = 0; % uncontact
                else
                    flag = 1; % contact 
                end
            end
        else
            flag = 0; % uncontact
        end
    end
    
end

File: jaxRBDL/Contact/DetectContact.py
import numpy as np
from jaxRBDL.Kinematics.CalcBodyToBaseCoordinates import CalcBodyToBaseCoordinates
from jaxRBDL.Kinematics.CalcPointVelocity import CalcPointVelocity
import jax.numpy as jnp

def DeterminContactType(pos: jnp.ndarray, vel: jnp.ndarray, contact_cond: dict)->int:
    pos = pos.flatten()
    vel = vel.flatten()
    contact_pos_lb = contact_cond["contact_pos_lb"].flatten()
    contact_vel_lb = contact_cond["contact_vel_lb"].flatten()
    contact_vel_ub = contact_cond["contact_vel_ub"].flatten()
    # print("------------")
    # print(pos)
    # print(vel)


    if pos[2] < contact_pos_lb[2]:
        if vel[2] < contact_vel_lb[2]:
            contact_type = 2 # impact    
        elif vel[2] > contact_vel_ub[2]:
            contact_type = 0 # uncontact
        else:
            contact_type = 1 # contact 
    else:
        contact_type = 0  # uncontact
    
    # if(contact_type !=0):
    #     print("------------")
    #     print(pos)
    #     print(vel)

    return contact_type

def  DetectContact(model: dict, q: jnp.ndarray, qdot: jnp.ndarray, contact_cond: dict)->jnp.ndarray:
    NC = int(model["NC"])

    try: 
        idcontact = jnp.squeeze(model["idcontact"], axis=0).astype(int)
        contactpoint = jnp.squeeze(model["contactpoint"], axis=0)
    except:
        idcontact = model["idcontact"]
        contactpoint = model["contactpoint"]

    flag_contact = jnp.zeros((NC, 1))


    flag_contact_list = []

    for i in range(NC):
        # Calcualte pos and vel of foot endpoint, column vector
        endpos_item = CalcBodyToBaseCoordinates(model, q, idcontact[i], contactpoint[i])
        endvel_item = CalcPointVelocity(model, q, qdot, idcontact[i], contactpoint[i])

        # Detect contact
        flag_contact_list.append(DeterminContactType(endpos_item, endvel_item, contact_cond))

    flag_contact = jnp.asarray(flag_contact_list).reshape((-1, 1))

    return flag_contact
    
